rezervo/integrations/fsc/booking.py

```diff
@@ -0,0 +1,229 @@
+            print(f"Found class, but weekday did not match: {c}")
+            result = BookingError.MISSING_SCHEDULE_DAY
+            continue
+        search_feedback = f'Found class: "{c.name}"'
+        if len(c.instructors) > 0:
+            search_feedback += (
+                f" with {format_name_list_to_natural([i.name for i in c.instructors])}"
+            )
+        else:
+            search_feedback += " (missing instructor)"
+        search_feedback += f" at {c.duration.start}"
+        print(search_feedback)
+        return rezervo_class_from_fsc_class(c)
+    err.log("Could not find class matching criteria")
+    if result is None:
+        result = BookingError.CLASS_MISSING
+    return result
+
+
+def book_fsc_class(auth_session: Session, class_id: int) -> bool:
+    print(f"Booking class {class_id}")
+    response = auth_session.post(
+        booking_url(auth_session),
+        json.dumps({"groupActivity": class_id, "allowWaitingList": True}),
+        headers={"Content-Type": "application/json"},
+    )
+    if response.status_code != 201:
+        err.log("Booking attempt failed: " + response.text)
+        return False
+    return True
+
+
+def book_class(
+    integration_user: IntegrationUser, _class: RezervoClass, config: ConfigValue
+) -> Union[None, BookingError, AuthenticationError]:
+    max_attempts = config.booking.max_attempts
+    if max_attempts < 1:
+        err.log("Max booking attempts should be a positive number")
+        return BookingError.INVALID_CONFIG
+    print("Authenticating...")
+    auth_session = authenticate_session(
+        integration_user.username, integration_user.password
+    )
+    if isinstance(auth_session, AuthenticationError):
+        err.log("Authentication failed")
+        return auth_session
+    booked = False
+    attempts = 0
+    while not booked:
+        booked = book_fsc_class(auth_session, _class.id)
+        attempts += 1
+        if booked:
+            break
+        if attempts >= max_attempts:
+            break
+        sleep_seconds = 2**attempts
+        print(f"Exponential backoff, retrying in {sleep_seconds} seconds...")
+        time.sleep(sleep_seconds)
+    if not booked:
+        err.log(
+            f"Booking failed after {attempts} attempt" + ("s" if attempts != 1 else "")
+        )
+        return BookingError.ERROR
+    print(
+        "Successfully booked class"
+        + (f" after {attempts} attempts!" if attempts != 1 else "!")
+    )
+    if config.notifications:
+        ical_url = "not supported"
+        notify_booking(config.notifications, _class, ical_url)
+    return None
+
+
+def cancel_fsc_booking(
+    auth_session: Session, booking_reference: int, booking_type: BookingType
+) -> bool:
+    print(f"Cancelling booking of class {booking_reference}")
+    res = auth_session.delete(
+        f"{booking_url(auth_session)}/{booking_reference}",
+        data=json.dumps({"bookingType": booking_type}),
```

Comment:
Is `json.dumps` required?

Reply:
Yup, same as above. The FSC API does not understand the request otherwise